profiler/profiler.go

```diff
@@ -0,0 +1,273 @@
+// criteria. If N <= 0, all the results based on the criteria are returned.
+func (p *Profiler) ReportTopNResults(numResults int, criteria []string) []ExprStats {
+	p.processLastExpr()
+	stats := []ExprStats{}
+
+	for _, hits := range p.hits {
+		for _, stat := range hits {
+			stats = append(stats, stat)
+		}
+	}
+
+	// allowed criteria for sorting results
+	allowedCriteria := map[string]lessFunc{}
+	allowedCriteria["EvalTime"] = func(stat1, stat2 *ExprStats) bool {
+		return stat1.ExprTimeNs > stat2.ExprTimeNs
+	}
+	allowedCriteria["NumEval"] = func(stat1, stat2 *ExprStats) bool {
+		return stat1.NumEval > stat2.NumEval
+	}
+	allowedCriteria["NumRedo"] = func(stat1, stat2 *ExprStats) bool {
+		return stat1.NumRedo > stat2.NumRedo
+	}
+
+	sortFuncs := []lessFunc{}
+
+	for _, cr := range criteria {
+		if fn, ok := allowedCriteria[cr]; ok {
+			sortFuncs = append(sortFuncs, fn)
+		}
+	}
+
+	// if no criteria return all the stats
+	if len(sortFuncs) == 0 {
+		return stats
+	}
+
+	orderedBy(sortFuncs).Sort(stats)
+
+	// if desired number of results to be returned is less than or
+	// equal to 0 or exceed total available results,
+	// return all the stats
+	if numResults <= 0 || numResults > len(stats) {
+		return stats
+	}
+	return stats[:numResults]
+
+}
+
+// Trace updates the profiler state.
+func (p *Profiler) Trace(event *topdown.Event) {
+	switch event.Op {
+	case topdown.EvalOp:
+		if expr, ok := event.Node.(*ast.Expr); ok && expr != nil {
+			p.processExpr(expr, event.Op)
+		}
+	case topdown.RedoOp:
+		if expr, ok := event.Node.(*ast.Expr); ok && expr != nil {
+			p.processExpr(expr, event.Op)
+		}
+	}
+}
+
+func (p *Profiler) processExpr(expr *ast.Expr, eventType topdown.Op) {
+
+	// set the active timer on the first expression
+	if p.activeTimer.IsZero() {
+		p.activeTimer = time.Now()
+		p.prevExpr = exprInfo{
+			op:       eventType,
+			location: expr.Location,
+		}
+		return
+	}
+
+	// record the profiler results for the previous expression
+	file := p.prevExpr.location.File
+	hits, ok := p.hits[file]
+	if !ok {
+		hits = map[Position]ExprStats{}
+		hits[Position{p.prevExpr.location.Row}] = getProfilerStats(p.prevExpr, p.activeTimer)
+		p.hits[file] = hits
+	} else {
+		pos := Position{p.prevExpr.location.Row}
+		pStats, ok := hits[pos]
+		if !ok {
+			hits[pos] = getProfilerStats(p.prevExpr, p.activeTimer)
+		} else {
+			pStats.ExprTimeNs += time.Since(p.activeTimer).Nanoseconds()
+
+			switch p.prevExpr.op {
+			case topdown.EvalOp:
+				pStats.NumEval++
+			case topdown.RedoOp:
+				pStats.NumRedo++
+			}
+			hits[pos] = pStats
+		}
+	}
+
+	// reset active timer and expression
+	p.activeTimer = time.Now()
+	p.prevExpr = exprInfo{
+		op:       eventType,
+		location: expr.Location,
+	}
+}
+
+func (p *Profiler) processLastExpr() {
+	expr := ast.Expr{
+		Location: p.prevExpr.location,
+	}
+	p.processExpr(&expr, p.prevExpr.op)
+}
+
+// Position represents a file location.
+type Position struct {
```

Comment:
It looks like this doesn't have to be exported. I would replace references to it simply with an `int`.